src/os/bluestore/BlueStore.cc

```diff
@@ -6515,16 +6525,30 @@ void BlueStore::_txc_finish_kv(TransContext *txc)
     finishers[n]->queue(txc->onreadable);
     txc->onreadable = NULL;
   }
-  if (txc->oncommit) {
-    finishers[n]->queue(txc->oncommit);
-    txc->oncommit = NULL;
-  }
-  while (!txc->oncommits.empty()) {
-    auto f = txc->oncommits.front();
-    finishers[n]->queue(f);
-    txc->oncommits.pop_front();
+  if (g_conf->bluestore_no_finisher && !txc->no_rtc &&
+      txc->osr->is_finisher_done()) {
+    if (txc->oncommit) {
+      txc->oncommit->complete(Context::NO_PGLOCK);
+      txc->oncommit = NULL;
+    }
+    while (!txc->oncommits.empty()) {
+      auto f = txc->oncommits.front();
+      f->complete(Context::NO_PGLOCK);
+      txc->oncommits.pop_front();
+    }
+  } else {
```

Comment:
This PG lock stuff isn't safe.  Even if it were, it should be separated out into a different patch.